### DISCUSSION TITLE
MddBootstrapInitialize :  ERROR_NOT_FOUND

### DISCUSSION THREAD

**Discussion by castorix** [microsoft/WindowsAppSDK#1352] [Q&A] [answered]
Hello,

I installed Microsoft.WindowsAppSDK.1.0-experimental1 (0.218.840.0)
from "WindowsAppSDKInstall.exe" 
but when I call MddBootstrapInitialize , I get :
hr = 2147943568 = 0x80070490  =ERROR_NOT_FOUND
Tested either in C++ or in C#
What could be the reason ?

Thanks in advance.

**Comment by riverar** [comment]
@castorix Can you show some code? What parameters do you pass into `MddBootstrapInitialize`? Did you install the correct WindowsAppSDKInstall for your target platform? (x86, x64, etc.)

**Reply by castorix** [reply]
> @castorix Can you show some code? What parameters do you pass into `MddBootstrapInitialize`? Did you install the correct WindowsAppSDKInstall for your target platform? (x86, x64, etc.)

In C++, I followed the code from [Tutorial: Build and deploy an unpackaged app that uses the Windows App SDK](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/tutorial-unpackaged-deployment#tabpanel_1_cpp)
(that I added in a Win32 Gui app, just after _wWinMain_)
But same behaviour with same parameters in C# (_MddBootstrapInitialize(1 << 16 | 0, "experimental1", 0);_)
I installed from "_1.0.0-Experimental1\Installer-x86\WindowsAppSDKInstall.exe_" from the extracted zip
I tested in Debug/x86 for both C++/C#

**Reply by riverar** [reply]
@castorix Can you provide the output of this PowerShell command? `Get-AppxPackage *WindowsAppSDK*`

**Reply by castorix** [reply]
> @castorix Can you provide the output of this PowerShell command? `Get-AppxPackage *WindowsAppSDK*`

I had checked it and it seems correct, except it displays x64 too, that was installed automatically with WindowsAppSDKInstall.exe x86 : 
[WindowsAppSDK](http://pastebin.fr/94962)

**Reply by riverar** [reply]
@castorix Is `Microsoft.WindowsAppSDK.Bootstrap.dll` getting copied into the output folder?

**Reply by castorix** [reply]
> @castorix Is `Microsoft.WindowsAppSDK.Bootstrap.dll` getting copied into the output folder?

Yes, otherwise I could not have been able to call the MddBootstrapInitialize API
(I also tried to call it dynamically in C++ (LoadLibrary/GetProcAddress) but same result

I just noticed that Microsoft.WindowsAppSDK.Main.1.0-experimental1 is only displayed in x64, not sure if it is normal...

**Reply by riverar** [reply]
And you're using the 1.0-experimental1 nuget as well?

Not seeing anything obvious here so adding @DrusTheAxe.

**Reply by riverar** [reply]
Here's a known working sample if you want to try it out in the meantime. https://github.com/riverar/mux-repros/tree/master/unpackaged/scratch

**Reply by castorix** [reply]
> And you're using the 1.0-experimental1 nuget as well?

Yes, it was the first thing I had installed...

**Reply by castorix** [reply]
> Here's a known working sample if you want to try it out in the meantime. https://github.com/riverar/mux-repros/tree/master/unpackaged/scratch

I had tried it and I get the same error...

**Reply by DrusTheAxe** [reply]
> > Here's a known working sample if you want to try it out in the meantime. https://github.com/riverar/mux-repros/tree/master/unpackaged/scratch
> 
> I had tried it and I get the same error...

What do you see when you run this at a normal (not admin) command line

```
powershell -c Get-AppxPackage *appsdk*
```

?

**Reply by castorix** [reply]
> What do you see when you run this at a normal (not admin) command line
> powershell -c Get-AppxPackage *appsdk*

It seems the same thing : 
[powershell -c Get-AppxPackage *appsdk*](http://pastebin.fr/94974)

**Reply by riverar** [reply]
What version/build of Windows are you on?

**Reply by castorix** [reply]
> What version/build of Windows are you on?

On Windows 10 1909,  18363.1621

**Comment by castorix** [comment]
I debugged by copying Microsoft.WindowsAppSDK.dll
 from C:\Program Files\WindowsApps into the exe directory

and it crashes in the function _FindDDLM_ at :
 _auto catalog{ winrt::Windows::ApplicationModel::AppExtensions::AppExtensionCatalog::Open(appExtensionName) };_
 with appExtensionName =  "microsoft.winappsdk.ddlm-1.0-x8-e1"
 (with version "1.0"  from majorMinorVersion)

**Reply by DrusTheAxe** [reply]
Merging threads...

>I just noticed that Microsoft.WindowsAppSDK.Main.1.0-experimental1 is only displayed in x64, not sure if it is normal...

Yes, if you have an x64 machine when WinAppSDK's installed you should have the followed registered:

* 2x Microsoft.WindowsAppSDK.1.0-experimental1 [x86 + x64]
* 1x Microsoft.WindowsAppSDK.Main.1.0-experimental1 [x64]
* 2x Microsoft.WindowsAppSDK.DDLM.1.0...x8-e1 and ...-x6-e1 [x86 + x64]

Framework packages can be installed side-by-side only differing by architecture. Thus an x64 system has both x64 + x86 flavors of the framework package.

Main packages can only have 1 package in the family registered for a user at a time. If there's multiple that _could_ work the most-applicable is chosen, thus same package family name, same version but different architecture the one matching the system is installed. Thus the x64 Main package on x64 systems.

The DDLM is main package (as in, PackageType=main) but we need SxS behavior, hence the reason their package name includes the architecture (x8=x86, x6=x64). That's the magic how we get both x86+x64 DDLM packages registered for a user - the package name differs so don't run afoul of the main-packages-can-only-have-1-package-in-the-family-registered-for-a-user-at-a-time restriction.

...
>with appExtensionName = "microsoft.winappsdk.ddlm-1.0-x8-e1"

Sounds like your app is an x86 process running on an x64 machine. Yes?

If so then your MddBoostrapInitialize call needs the following packages registered for the user (you can check via `powershell -c get-appxpackage *appsdk*`):

* x86 framework package (`Microsoft.WindowsAppSDK.1.0-experimental1`)
* x64 Main package (`Microsoft.WindowsAppSDK.Main.1.0-experimental1`)
* x86 DDLM package (`Microsoft.WindowsAppSDK.1.0-...x8-e1`)

> and it crashes in the function FindDDLM at :

Is it _crashing_ or _failing_? (sorry to be pedantic, but those words have very specific meaning here)

If the DDLM x8-e1 package isn't registered for the user then the AppExtensionCatalog::Open() call won't find a match and throw an exception, but it's the not-found error case and should boil up as a FAILED(hresult) return value from `MddBootstrapInitialize`. If that's the problem you can run the installer, or manually install the packages via

```
cd ...dir-where-you-have-the-nuget-unzipped-on-disk...\tools\MSIX
powershell -c add-appxpackage win10-x86\Microsoft.WindowsAppSDK.1.0.msix
powershell -c add-appxpackage win10-x86\Microsoft.WindowsAppSDK.DDLM.1.0.msix
powershell -c add-appxpackage win10-x64\Microsoft.WindowsAppSDK.1.0.msix
powershell -c add-appxpackage win10-x64\Microsoft.WindowsAppSDK.Main.1.0.msix
```

VS doesn't install those today when you hit F5 (something we're working on for 1.0 but not ready yet).


OTOH if there's an actual crash in there, can you share a callstack?

**Reply by castorix** [reply]
> If so then your MddBoostrapInitialize call needs the following packages registered for the user (you can check via powershell -c get-appxpackage *appsdk*):
> x86 DDLM package (Microsoft.WindowsAppSDK.1.0-...x8-e1)

I don't have 1.0, but the version 0.218.840 in the name :
Microsoft.WindowsAppSDK.DDLM.0.218.840.0-x8-e1

I tried
powershell -c remove-appxpackage Microsoft.WindowsAppSDK.DDLM.0.218.840.0-x8-e1_0.218.840.0_x86__8wekyb3d8bbwe
powershell -c add-appxpackage Microsoft.WindowsAppSDK.DDLM.1.0-experimental1.msix
then
powershell -c get-appxpackage *appsdk*
but I still get
Microsoft.WindowsAppSDK.DDLM.0.218.840.0-x8-e1

then in the code :

![FindDDLM](https://user-images.githubusercontent.com/22345506/132951783-da3ef509-afaa-4706-8f55-e34df79fe1c2.jpg)

which returns : 
hr = HRESULT_FROM_WIN32(ERROR_NOT_FOUND)

**Comment by castorix** [comment]
I removed all packages and installed the last preview2 version: 
Still same **ERROR_NOT_FOUND** error

_Deploying package: Microsoft.WindowsAppRuntime.1.0-preview2_0.272.1934.0_x64__8wekyb3d8bbwe
Package deployment result : 0x0
Deploying package: Microsoft.WindowsAppRuntime.1.0-preview2_0.272.1934.0_x86__8wekyb3d8bbwe
Package deployment result : 0x0
Deploying package: Microsoft.WindowsAppRuntime.Main.1.0-preview2_0.272.1934.0_x64__8wekyb3d8bbwe
Package deployment result : 0x0
Provisioning result : 0x0
Deploying package: Microsoft.WindowsAppRuntime.Singleton-preview2_0.272.1934.0_x64__8wekyb3d8bbwe
Package deployment result : 0x0
Provisioning result : 0x0
Deploying package: Microsoft.WinAppRuntime.DDLM.0.272.1934.0-x6-p2_0.272.1934.0_x64__8wekyb3d8bbwe
Package deployment result : 0x0
Provisioning result : 0x0
Deploying package: Microsoft.WinAppRuntime.DDLM.0.272.1934.0-x8-p2_0.272.1934.0_x86__8wekyb3d8bbwe
Package deployment result : 0x0
Provisioning result : 0x0
All packages were installed successfully._

**Reply by DrusTheAxe** [reply]
OK, 1.0-preview2 should be installed correctly. Can you share your call to MddBootstrapInitialize()? Its majorMinor parameter should be 0x0001000 and its versionTag="preview2"

**Reply by castorix** [reply]
It is 0x00010000, not 0x0001000  =>

```
// Take a dependency on Windows App SDK 1.0 Preview 1. If using 1.0 Experimental, 
// replace with versionTag{ L"experimental1" }. If using version 0.8 Preview, 
// replace with majorMinorVersion{ 0x00000008 } and  versionTag{ L"preview" }.
const UINT32 majorMinorVersion{ 0x00010000 };
PCWSTR versionTag{ L"preview2" };
const PACKAGE_VERSION minVersion{};

const HRESULT hr{ MddBootstrapInitialize(majorMinorVersion, versionTag, minVersion) };
```

![MddBootstrapInitialize_bug](https://i.ibb.co/9p28Bt8/Mdd-Bootstrap-Initialize-bug2.jpg)

**Reply by DrusTheAxe** [reply]
Aha! I think I see the problem! You need to initialize COM on the thread before calling the bootstrap API

Try adding this as line 27.5

```
const HRESULT hrCoInit{ ::CoInitializeEx(nullptr, COINITBASE_MULTITHREADED) };
if (FAILED(hrCoInit))
{
    printf("Oh no! 0x%X", hrCoInit);
    return hrCoInit;
}
```

**Reply by castorix** [reply]
> const HRESULT hrCoInit{ ::CoInitializeEx(nullptr, COINITBASE_MULTITHREADED) };

No, it doesn't change...

And I have the same bug with the C# sample from [scratch](https://github.com/riverar/mux-repros/tree/master/unpackaged/scratch)

**Reply by riverar** [reply]
What OS version/build are you on? Is this a machine that had any privacy/"debloat" scripts run on?

**Reply by castorix** [reply]
> What OS version/build are you on? Is this a machine that had any privacy/"debloat" scripts run on?

It is a standard Windows 10 Education, Version 1909, Build 18363.1801

**Reply by DrusTheAxe** [reply]
> > What OS version/build are you on? Is this a machine that had any privacy/"debloat" scripts run on?
> 
> It is a standard Windows 10 Education, Version 1909, Build 18363.1801

TL;DR known bug in the bootstrapper on &lt;20H1. The fix was just checked in (Unpackaged Windows App SDK app doesn't run on RS5/19H1 #1681) to cook through the overall build pipeline and verification. The next 1.0 release should have the fix, so, soon.

Per my [handy dandy decoder ring](https://en.wikipedia.org/wiki/Windows_10_version_history#Channels) Windows 10 Version 1909 == 19H2, but 19H2 has the same limitation for the bootstrap as 19H1 (no surprise, as 19H2 is (oversimplified) 19H1 plus some fixes).

**Comment by pjc50** [comment]
I'm having the same problem, except I definitely have the version as 0x10000 from the example code at https://github.com/MicrosoftDocs/windows-uwp/issues/3439

I notice that installing the runtime gives packages named Microsoft.WindowsAppRuntime.DDLM.1.0-preview3.msix etc whereas all the discussion above refers to Microsoft.WindowsAppSDK.DDLM - is this a different package or has it been renamed? Looks like it's been renamed given https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/check-windows-app-sdk-versions

**Reply by DrusTheAxe** [reply]
Do you initialize COM on the thread before calling the bootstrap API? See https://github.com/microsoft/WindowsAppSDK/discussions/1352#discussioncomment-1555921

> packages named Microsoft.WindowsAppRuntime.DDLM.1.0-preview3.msix

Correction: ...package _files_ named...

A "packaged named..." is usually referring to the package's _identity_, which is defined by `<Identity>` in its `appxmanifest.xml`. Filenames have no bearing on package identity.

Package identities are more detailed (and don't end in ".msix" :P). See [MSIXPackages.md](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/Deployment/MSIXPackages.md) for more details.

>discussion above refers to Microsoft.WindowsAppSDK.DDLM

A common term for DDLM package(s). A DDLM package's Name, as in 'the Name field in its package identity' is spec'd [here](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/Deployment/MSIXPackages.md#33-package-naming---ddlm)

* Name = Microsoft.WinAppRuntime.DDLM.<Version>-<ShortArchitecture>[-ShortVersionTag]

In practice that means package Names like `Microsoft.WinAppRuntime.DDLM.0.297.2018.0-x6-p3`

and a Package Full Name like `Microsoft.WinAppRuntime.DDLM.0.297.2018.0-x6-p3_0.297.2018.0_x64__8wekyb3d8bbwe`

>is this a different package or has it been renamed?

DDLM packages are rather unusual, as they have a unique Name for every version+architecture (see above). This is a creative trick to enable Dynamic Dependencies to let unpackaged apps use framework packages (without changing Windows). See [MSIXPackages.md](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/Deployment/MSIXPackages.md) and [DynamicDependencies.md](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/dynamicdependencies/DynamicDependencies.md) for more details

P.S. "package Name" above is not the same as "package name". The latter is a rather imprecise term. The former refers to the Name field in a package's identity. Package identity is specific concept, as a combination of...

* Name
* Version
* Archictecture
* ResourceId
* Publisher

and can be expressed with multiple syntax including [struct PACKAGE_INFO](https://docs.microsoft.com/en-us/windows/win32/api/appmodel/ns-appmodel-package_info), [Windows.ApplicationModel.PackageId](https://docs.microsoft.com/en-us/uwp/api/Windows.ApplicationModel.PackageId?view=winrt-22000) and more but, most commonly, the 5-tuple (the fields noted above) or as a string (aka Package Full Name).

There's lots of documentation, APIs, blog posts and more referencing package identity but I'm not sure where it's formally defined.

**Reply by pjc50** [reply]
I have not explicitly initialised COM. I was under the impression that the CLR does this for the main thread? The entire program (taken from the Microsoft documentation):

```
using System;
using Microsoft.Windows.ApplicationModel.DynamicDependency;

namespace DynamicDependenciesTest
{
    class Program
    {
        [STAThread]
        static void Main(string[] args)
        {
            Bootstrap.Initialize(0x00010000, "preview3");
            Console.WriteLine("Hello World!");
            // Release the DDLM and clean up.
            Bootstrap.Shutdown();
        }
    }
}
```
I have tried with and without [STAThread]. Calling CoInitializeEx((System.IntPtr)null, COINIT.COINIT_APARTMENTTHREADED) returns S_FALSE and doesn't help the subsequent failure.

**Reply by pjc50** [reply]
The C++ example from
https://github.com/MicrosoftDocs/windows-uwp/blob/docs/hub/apps/windows-app-sdk/tutorial-unpackaged-deployment.md does not initialize COM. It also doesn't work for me, and likewise calling CoInitializeEx does not make it work.

**Reply by pjc50** [reply]
My minimal C++ testcase is now
```
#pragma comment(lib, "windowsapp")
#include <winrt/base.h>
#include <winrt/Windows.Foundation.h>
#include <winrt/Windows.ApplicationModel.AppExtensions.h>

int main()
{
    winrt::init_apartment();
    auto catalog{ winrt::Windows::ApplicationModel::AppExtensions::AppExtensionCatalog::Open(L"microsoft.winappruntime.ddlm-1.0-x6-p3") };
}
```
Having isolated the failing call and brought it up where I can see it. It looks like the problem is the appextension winrt interface itself failing to load, but I'm at a loss as to why. System is Windows 10 Version 1909, "10.0.18363.1801". Project target version is "10.0.19041.0" (but I've also tried 17763. All within VS2019.

**Reply by pjc50** [reply]
This is the same issue as https://github.com/microsoft/WindowsAppSDK/issues/1121#issuecomment-934640347 , except not on Windows Server. The app extension API is .. missing? Which makes it unavailable to the bootstrapper.

**Reply by castorix** [reply]
> This is the same issue as [#1121 (comment)](https://github.com/microsoft/WindowsAppSDK/issues/1121#issuecomment-934640347) , except not on Windows Server. The app extension API is .. missing? Which makes it unavailable to the bootstrapper.

I finally decided to upgrade my Windows version from 1909 to 21H1... and now it works !
(tested in C++ and C#)
So there is clearly a problem with Windows 10 1909, 18363.1801

**Reply by DrusTheAxe** [reply]
Yes. There was a Windows limitation we weren't aware of in &lt;20H1 causing the Bootstrap API to fail. The fix changes the Bootstrap's implementation to avoid this problem.

For more details see Unpackaged Windows App SDK app doesn't run on RS5/19H1 #1681

**Comment by castorix** [comment, marked as answer]
I mark the thread as answered

from comments :
https://github.com/microsoft/WindowsAppSDK/discussions/1352#discussioncomment-1573928
https://github.com/microsoft/WindowsAppSDK/discussions/1352#discussioncomment-1559521

**Comment by triplef** [comment]
I’m getting the same error `0x80070490` from `MddBootstrapInitialize2()` with the latest SDK version 1.5.3 (1.5.240428000 / 5001.119.156.0) on Windows 11 23H2.

I already tried a full uninstall (`get-appxpackage micro*win*appruntime* | remove-appxpackage`) and reinstall with the same result.

Any idea @DrusTheAxe?

**Reply by DrusTheAxe** [reply]
0x80070490 means WinAppSDK isn't registered for the user as you expect. This is usually because all the needed packages didn't register, or your process isn't running as the user you think it is (or both).

You can do a few easy things to investigate:

* For the process calling the Bootstrapper...
  * What's the user context?
  * What's the security context? (Elevated, MediumIL, AppContainer, etc)
* Are the packages registered for the user?
  * At an admin prompt run `powershell -c get-appxpackage micro*win*app*run* -AllUsers
    * The `PackageUserInformation` property will show a list of `usersid [username]: status`
      * Status of 'Installed' means it's registered for that user
      * Status of 'Staged' means it's staged on the machine.
        * NOTE: 'Staged` should only appear for LocalSystem (S-1-5-18)
        * NOTE: A package should be `Staged` for LocalSystem -OR- `Installed (registered) for 1 or more users. If both then something's wrong
  * Are all the packages registered for the user?
    * NOTE: The Framework (Microsoft.WindowsAppRuntime.1.5*) and DDLM packages are critical for the Bootstrapper
* Was there a Deployment problem installing the packages?
  * Review the EventLog channels `Microsoft-Windows-AppXDeploymentServer/Operational` and `Microsoft-Windows-AppXDeployment/Operational` e.g. run eventvwr.exe and drill down...
![image](https://github.com/microsoft/WindowsAppSDK/assets/72222/fb905660-2ee8-4c8f-a8d6-82649661ef8e)
  * Every time Deployment processes an MSIX package you'll find events about a request to process the package, 1 or more events about the processing and an Informational event at the end about the request succeeding or an Error if it failed. If nothing's logged it's because the request never reached the Deployment pipeline, in which case the `Microsoft-Windows-AppXDeployment/Operational` channel should have more info (the [PackageManager APIs](https://learn.microsoft.com/uwp/api/windows.management.deployment.packagemanager?view=winrt-22621) log here before/after handing the request to the Deployment pipeline)
